[16.0][WIP][ADD] account_liquidity_forecast

JordiBForgeFlow opened the pull request:
Module presented during OCA Days https://odoo-community.org/event/oca-days-2023-liege-2023-11-06-2023-11-07-143/track/02-45-treasury-management-let-s-make-it-happen-217

Slides: https://docs.google.com/presentation/d/1fSRMP9tK_YQCA6sQFS9mzXKVDmOLAIBhOSojxfA_AF4/edit?usp=sharing

**This is work in progress**

Todo:
* [x] Export to PDF
* [x] Export to XLSX
* [ ] Export to Odoo Spreadsheets ¿?
* [ ] Add tests

JordiBForgeFlow committed:
[ADD] account_liquidity_forecast

gjotten commented:
I would posit that exporting to spreadsheets might be a nice feature to add through an extension module?

andreampiovesana commented:
@JordiBForgeFlow news?

andreampiovesana commented:
if you have empty db and you try, 
"/home/odoo-dev/odoo-dev16/PR/account_liquidity_forecast/report/liquidity_forecast.py", line 66, in _complete_beginning_balance
    initial_balance = initial_balances[0]
IndexError: list index out of range
should control without initial balance

JordiBForgeFlow commented:
@andreampiovesana Fixed! Please try again

andreampiovesana commented:
> @andreampiovesana Fixed! Please try again

ok now, EXPORT XLSX no output, VIEW and ESPORT PDF yes

JasminSForgeFlow committed:
[IMP] account_liquidity_forecast: consider partial payments when deciding non-reconciled items in a given period

JordiBForgeFlow committed:
[FIX] account_liquidity_forecast: consider only bank journals to
get the incoming and outgoing payment accounts

JordiBForgeFlow committed:
[FIX] account_liquidity_forecast: fix filtering of open items

JasminSForgeFlow committed:
[IMP] account_liquidity_forecast: Add XLSX report

JasminSForgeFlow committed:
[IMP] account_liquidity_forecast: Improvements in report formatting

rvalyi commented on account_liquidity_forecast/reports.xml:
```diff
@@ -7,7 +7,7 @@
         <field name="page_height">297</field>
         <field name="page_width">210</field>
-        <field name="orientation">Portrait</field>
+        <field name="orientation">Landscap</field>
```
shouldn't that be Landscape with an e?

JasminSForgeFlow replied:
its updated,

Thanks

MateuGForgeFlow replied:
@rvalyi your request is added

Can you please review again?

Thank you!

JordiBForgeFlow committed:
[IMP] account_liquidity_forecast: consider open items at current date
With previous approach the results were incorrect. We would be proposing
to show the items that are open back in the bucket date, but in reality
we want to see, based on today, what is open based on the due date of the
open item.

andreampiovesana commented:
@JordiBForgeFlow can we merge as draft?

MaurizioPellegrinet reviewed (approved):
Functional test: LGTM

JordiBForgeFlow commented:
@MaurizioPellegrinet @andreampiovesana Can you add some tests?

MaurizioPellegrinet commented:
@JordiBForgeFlow It's possible to integrate:
- sales order forecast
- supplier order forecast
- add accounts or account groups

JordiBForgeFlow commented:
@MaurizioPellegrinet You can make a proposal. I suggest as a separate module. In the meanwhile I suggest that you contribute to add some tests to the module, as it is.

MaurizioPellegrinet commented:
@JordiBForgeFlow here some tests

[Liquidity forecast.odt](https://github.com/user-attachments/files/16512065/Liquidity.forecast.odt)

JordiBForgeFlow commented:
@MaurizioPellegrinet I mean tests in the code.

MaurizioPellegrinet commented:
@JordiBForgeFlow sorry, I'm functional (only)

MaurizioPellegrinet commented:
@JordiBForgeFlow Hi, in Liquidity Forecast Planning Groups is possible to duplicate the row or import list to .xlsx?
These options is very important if I have many repeating movements

douglascstd commented:
Dear Friends,

Great job on the liquidity forecast module; it's easy to create the report.

I'd like to present some modules that offer a similar purpose but from a different perspective. These modules use the MIS Builder Cashflow as the base and are designed to populate the mis.cash_flow.forecast_line model with source data calculated in each module.

### base:
[mis_builder_cash_flow_forecast_integration](https://github.com/Escodoo/account-addons/tree/14.0/mis_builder_cash_flow_forecast_integration)
### Sources:
[mis_builder_cash_flow_forecast_purchase](https://github.com/Escodoo/account-addons/tree/14.0/mis_builder_cash_flow_forecast_purchase)
[mis_builder_cash_flow_forecast_sale](https://github.com/Escodoo/account-addons/tree/14.0/mis_builder_cash_flow_forecast_sale)
[mis_builder_cash_flow_forecast_contract](https://github.com/Escodoo/account-addons/tree/14.0/mis_builder_cash_flow_forecast_contract)
[mis_builder_cash_flow_forecast_sale_invoice_plan](https://github.com/Escodoo/account-addons/tree/14.0/mis_builder_cash_flow_forecast_sale_invoice_plan)

In the future, consider adding:
    Sale Blanket Contract

All of these modules are currently in use with some customer instances but are not yet available in the OCA environment. I am confident that the author @marcelsavegnago aims to include them there.

I don’t want to disrupt the current work but merely wanted to present an alternative approach to achieving the same goal.

andreampiovesana commented:
No close, i use it

vdewulf commented:
Hello @JordiBForgeFlow
I am on the Board of a Food Hub cooperative and we are working on our Liquidity Forecast. The MIS Builder Cashflow report is nice but doesn't allow to adapt the realistic expected dates of payement (we know that 20% of the invoices are paid after the due date). 
Would this solution help on this specific topic?
And also I am wondering what can be done for the next steps on this PR: missing tests in code, right?
Thanks!

MateuGForgeFlow committed:
[16.0][FIX] account_liquidity_forecast: correct use of move state

MateuGForgeFlow committed:
[16.0][IMP] account_liquidity_forecast: improve readme

andreampiovesana commented:
merge?

MateuGForgeFlow committed:
[16.0][IMP] account_liquidity_forecast: improve date format

MateuGForgeFlow committed:
[16.0][IMP] account_liquidity_forecast: add tests

MateuGForgeFlow committed:
[16.0][IMP] account_liquidity_forecast: date format company week start

MateuGForgeFlow committed:
[16.0][IMP] account_liquidity_forecast: compute column width dynamically

MateuGForgeFlow commented:
Main Changes:
- Improved report column width computation.
- Improved week format.
- Corrected draft entry feature.
- Added tests.

@andreampiovesana let's see if we get approval so we can merge!

Thank you!

MateuGForgeFlow reviewed (approved):
Functional and Code review

andreampiovesana reviewed (approved):
LGTM

andreampiovesana commented:
@rvalyi do you approve?

andreampiovesana commented:
merge?

rvalyi commented:
as per reviews,

ocabot merge nobump

rvalyi commented:
as per reviews,

/ocabot merge nobump
